### PR TITLE
Compute Centroids from CSV

### PR DESCRIPTION
**Pull request by GNiendorf**
This PR adds a compute_centroids_csv.py file that uses one of the csv files contained in data/ to compute the sensor centroid coordinates. I also deleted all of the old files in data/ since it wasn't documented which event files these came from and they are very old/outdated.

### CODE REVIEW

**Comment by slava77** [comment]
https://github.com/cms-sw/cmssw/blob/master/RecoTracker/MkFit/test/dumpMkFitGeometryPhase2.py
can be used to make a bin file 

The code to read the mkfit TrackerInfo is in 
https://github.com/cms-sw/cmssw/blob/e9815c62b4a396647ccffb9396f70b8c458ee576/RecoTracker/MkFitCore/src/TrackerInfo.cc#L202

**Comment by GNiendorf** [comment]
coord_diff

Here's a plot of the differences between the current method and using the geometry csv file.

**Comment by GNiendorf** [comment]
Sorry, converted this back to a draft PR. I saw that there are some moduletype differences. I need to check my functions that determine the module type and find the mistake.

**Comment by GNiendorf** [comment]
Below I'm showing the difference between the moduletype (23, 24, 25) for the old method vs the csv method on all sensors. There is now perfect agreement after the last commit.

modtype_diff

**Comment by slava77** [comment]
> Here's a plot of the differences between the current method and using the geometry csv file.

I was expecting microns, not mm.
At least in the direction normal to the module surface it should be microns.

**Comment by GNiendorf** [comment]
> I was expecting microns, not mm. At least in the direction normal to the module surface it should be microns.

Is approximating the center of each sensor using hits expected to be super accurate? I guess if you can make the assumption that hits land on the sensors in random positions such that their average isn't biased.

**Comment by slava77** [comment]
> > At least in the direction normal to the module surface it should be microns.
> 
> Is approximating the center of each sensor using hits expected to be super accurate?

in the direction normal to the module surface it has to be "super accurate"

**Comment by GNiendorf** [comment]
Here's a plot of the total distance as well.

total_dist

File: compute_centroid_csv.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    det_id (int): The detector ID of the sensor.
+
+    Returns:
+    int: The numerical module type. 
+         -1 for inner tracker modules,
+         23 (PSP), 24 (PSS), or 25 (TwoS) for different module types.
+    
+    Note: 
+    For more information, see
+    https://github.com/cms-sw/cmssw/tree/master/Geometry/TrackerGeometryBuilder
```
Comment (VourMa) [inline compute_centroid_csv.py:34]: Probably add also the caveat that this could/should be improved (become more robust) in the future?

**Review by VourMa — COMMENTED** [review]
The PR looks good to me. Before I approve and merge, a couple of question to understand the strategy:
- Don't we need an update in the README to document where/how this newly created script fits in the whole workflow?
- Are we done with the tests/plots that @slava77 had requested?
- By the comments, it seems like the differences in module positions are larger than expected. This may mean that there could be non-negligible differences downstream. I understand that the product of these updates are meant to be used in the actual LST code, so why don't we start testing it already? (I recall that there was an argument not to in the last meeting but I don't remember exactly, sorry for this.)

**Comment by GNiendorf** [comment]
Thank you for the feedback Manos, I'll work on the changes you suggested today. As for the comparison plots, the orientation scripts still make use of the hit information. I'm not sure if using the exact centroids from the csv file combined the orientations from the hit info will make for a meaningful evaluation. I think therefore it may just be better to wait until the later files also use the csv file info before running the full plots. What do you think?

**Comment by VourMa** [comment]
>  I'm not sure if using the exact centroids from the csv file combined the orientations from the hit info will make for a meaningful evaluation. I think therefore it may just be better to wait until the later files also use the csv file info before running the full plots. What do you think?

Up till now we went with the assumption that the hit information is close to the truth and good for us to work with. If that assumption is true, we should see negligible differences in the validation of this PR. If not, then we will see changes in the performance. In any case, a full validation would give us a sense of how correctly we are doing things and this is useful info, so I would still do it.

Just to understand the plan though: are you planing to update also the orientations with the csv file info but in another PR? If so, how extensive (in developments and in time) of an update that will be. Same magnitude as this one?

File: data/DetId_sensors_list.csv
[the remark this answers is not in the document]
Comment (GNiendorf) [reply data/DetId_sensors_list.csv:1]: Thanks, fixed in the commit [414e322](https://github.com/SegmentLinking/LSTGeometry/pull/8/commits/414e32258a97d7e07498108d2f690c1f10928863)

File: compute_centroid_csv.py (hunk 1 above)
Comment (GNiendorf) [reply compute_centroid_csv.py]: Fixed in the commit [414e322](https://github.com/SegmentLinking/LSTGeometry/pull/8/commits/414e32258a97d7e07498108d2f690c1f10928863)

File: README.md [hunk 2]
```diff
@@ -29,6 +29,16 @@ Take a PU200 ttbar sample so that we have enough hits to perform some simple fit
     # remember the port number and do something like the following on your local computer
     ssh -N -f -L localhost:$port:localhost:$port uaf-$uaf.t2.ucsd.edu
 
+## Geometry File Information
+
+The files `DetId_sensors_list.csv` and `module_info.csv` are sourced from Tracker version OT806_IT741:
+
+- URL: [Tracker Version OT806_IT741](https://cms-tklayout.web.cern.ch/cms-tklayout/layouts-work/recent-layouts/OT806_IT741/info.html)
```
Comment (VourMa) [inline README.md:36]: From [here](https://cms-tklayout.web.cern.ch/cms-tklayout/layouts-work/recent-layouts/index.html), I see that the layout used is actually based on T30 with additional updates - do these correspond to a CMSSW tracker version, I am not sure.
image

However, based on our [sample generation wiki](https://github.com/SegmentLinking/TrackLooper/wiki/Line-Segment-Tracking-Sample-Generation), our samples are based on `CMSSW_12_6_3`. Looking CMSSW to find the [default geometry back then (bottom of the page)](https://github.com/cms-sw/cmssw/tree/CMSSW_12_6_3/Configuration/Geometry) then, I see T24.

From [here](https://cms-tklayout.web.cern.ch/cms-tklayout/layouts-work/cmssw-models/index.html), I see that T24 was based on [OT800_IT615](https://cms-tklayout.web.cern.ch/cms-tklayout/layouts-work/cmssw-models/YB_OT800_IT615/index.html). Isn't this one the one we should be using? How was the specific layout used in this PR chosen?

Having said that, I was looking through the evolution of the tracker layouts, and the differences in the OT should be minimal, so we may still be ok. But we try to ensure that we are consistent.

By the way, we should be soon updating to a new geometry with new samples as well: [new CMSSW geometry default](https://github.com/cms-sw/cmssw/tree/master/Configuration/Geometry).

Comment (GNiendorf) [reply README.md:36]: Nice catch, we forgot to consider that our samples use an older geometry when choosing this version. I'll fix this right now and change it to OT800_IT615 like you mentioned.

Comment (GNiendorf) [reply README.md]: geo

Here's a set of comparison histograms between the old centroids with the OT806_IT741 geometry and the OT800_IT615 geometry. I'll add some more plots that compare the differences between the hit method vs this new CSV method with the updated geometry and the plots that Slava requested.

**Comment by GNiendorf** [comment]
Here's an updated comparison of the current hit-based method vs. CSV method using the corrected geometry CSV file mentioned in the discussion/commit above. The asymmetry in the Z direction went away which looks promising. I'm going to make a second plot like discussed before along the plane perpendicular to the modules in the barrel to make sure they match exactly. After that I'll run the full set of plots for reference and we can merge this in.

old_vs_new

**Comment by GNiendorf** [comment]
@VourMa Comparison plots [Here](https://www.classe.cornell.edu/~gsn27/www/www/PRLST/mywork_50d8a2-PU200_50d8a2D-PU200/summary/)

The plots above are using the serial CPU implementation, so any variations you see are coming from the module maps. Differences are very small, but seems like a slight improvement overall? I think this PR is ready to be merged in either way.

**Review by VourMa — APPROVED** [review]
The validation looks ok and all of the comments have been resolved. I am approving and merging this, thanks @GNiendorf!

**Pull request merged**